File: Codeforces/Rounds/VK Cup/2022/Qual (Engine)/D2_Generator.py
import random


def read_int():
    return int(input())


def read_ints():
    return map(int, input().split())


def solve():
    values = "6789TJQKA"
    modes = "CDSH"

    nine = values.index('9')
    players_count = 2

    all_cards = []
    for mode in range(len(modes)):
        for value in range(len(values)):
            all_cards.append((value - nine, mode))

    hand_size = len(all_cards) // players_count

    def get_ans(hands):
        left_size = nine + 1
        right_size = len(values) - nine

        empty = left_size * right_size
        total = empty + 1

        def segments_to_state(segments):
            state = 0
            for segment in segments:
                segment_value = empty

                if segment is not None:
                    left, right = segment
                    segment_value = abs(left) * right_size + abs(right)

                state *= total
                state += segment_value

            return state

        best_result = [dict() for _ in range(players_count)]

        def calc_result(segments, player, waiting_counts):
            state = segments_to_state(segments)
            if state in best_result[player]:
                return best_result[player][state]

            possible = []
            waiting_count = waiting_counts[player]

            hand = hands[player]
            for card in hand:
                value, mode = card
                if segments[mode] is None:
                    if value == 0:
                        possible.append(card)
                else:
                    left, right = segments[mode]
                    if value == left - 1 or value == right + 1:
                        possible.append(card)

            other_player = 1 - player

            if len(possible) == 0:
                best_result[player][state] = calc_result(segments, other_player, waiting_counts)
            elif waiting_count == 1:
                waiting_counts[player] -= 1
                best_result[player][state] = waiting_counts[1] - waiting_counts[0]
                waiting_counts[player] += 1
            else:
                best_result[player][state] = hand_size * (-1 if player == 0 else 1)
                player_func = max if player == 0 else min

                for value, mode in possible:
                    mode_segment = segments[mode]

                    if value == 0:
                        segments[mode] = (0, 0)
                    else:
                        left, right = mode_segment
                        segments[mode] = (min(left, value), max(right, value))

                    waiting_counts[player] -= 1
                    result = calc_result(segments, other_player, waiting_counts)
                    waiting_counts[player] += 1

                    segments[mode] = mode_segment

                    best_result[player][state] = player_func(best_result[player][state], result)

            return best_result[player][state]

        alice, bob = 0, 1

        start = [None] * 4
        start_waiting = [hand_size, hand_size]
        alice_result = calc_result(start, alice, start_waiting)
        bob_result = calc_result(start, bob, start_waiting)

        return abs(alice_result - bob_result)

    def card_to_str(card):
        value, mode = card
        return values[value + nine] + modes[mode]

    need_size = read_int()

    results_to_hands = dict()
    while len(results_to_hands) < need_size:
        random.shuffle(all_cards)

        gen_hands = [
            all_cards[player * hand_size:(player + 1) * hand_size]
            for player in range(players_count)
        ]

        result = get_ans(gen_hands)
        if result not in results_to_hands:
            results_to_hands[result] = [
                list(map(card_to_str, hand))
                for hand in gen_hands
            ]

    ans_hands_list = list(results_to_hands.values())[:need_size]
    for ans_hands in ans_hands_list:
        for ans_hand in ans_hands:
            print(*ans_hand)
        print()


if __name__ == '__main__':
    solve()

File: Codeforces/Rounds/VK Cup/2022/Qual (Engine)/D2_Generator.cpp
#include <bits/stdc++.h>

using namespace std;

using ll = int64_t;
using vi = vector<int>;
using ii = pair<int, int>;
using vii = vector<ii>;

void solve_test(int need_size) {	
	string values = "6789TJQKA";
	string modes = "CDSH";
	
    	int nine = values.find('9');
	int players_count = 2;
	
	vii all_cards;
    	for (int mode = 0; mode < modes.size(); ++mode) {
    		for (int value = 0; value < values.size(); ++value) {
    			all_cards.emplace_back(value - nine, mode);
    		}
    	}

	int hand_size = all_cards.size() / players_count;
	
	int left_size = nine + 1;
        int right_size = values.size() - nine;

	int empty = left_size * right_size;
	ii empty_segment = make_pair(1, -1);
	
	int total = empty + 1;

	auto segments_to_state = [&](vii& segments) {
            int state = 0;
            for (auto [left, right] : segments) {
                int segment_value = empty;

                if (left <= 0 && right >= 0) {
                    segment_value = abs(left) * right_size + abs(right);
                }

                state *= total;
                state += segment_value;
            }

            return state;
        };
	
	auto get_ans = [&]() {
		vii segments(modes.size(), empty_segment);
		vi waiting_counts(players_count, hand_size);
	
		using cache_type = vector<int>;
		
		int max_state = segments_to_state(segments) + 1;
		int none = -100500;
		vector<cache_type> best_result(players_count, cache_type(max_state, none));
		
		vector<vi> card_to_index(modes.size(), vi(values.size()));
		for (int card_index = 0; card_index < all_cards.size(); ++card_index) {
			auto& [value, mode] = all_cards[card_index];
			card_to_index[mode][value + nine] = card_index;
		}
		
		function<int(int)> calc_result = [&](int player) {
	            int state = segments_to_state(segments);
		    if (best_result[player][state] != none) {
	                return best_result[player][state];
	            }

		    vi possible;

		    for (int mode = 0; mode < modes.size(); ++mode) {
		    	if (segments[mode] == empty_segment) {
		    		int nine_index = card_to_index[mode][nine];
		    		if (nine_index / hand_size == player) {
		    			possible.push_back(nine_index);
		    		}
		    	} else {
		    		auto& [left, right] = segments[mode];
		    		
		    		int next_left = left - 1;
		    		if (next_left + nine >= 0) {
		    			int card_index = card_to_index[mode][next_left + nine];
		    			if (card_index / hand_size == player) {
		    				possible.push_back(card_index);
		    			}
		    		}
		    		
		    		int next_right = right + 1;
		    		if (next_right + nine < values.size()) {
		    			int card_index = card_to_index[mode][next_right + nine];
		    			if (card_index / hand_size == player) {
		    				possible.push_back(card_index);
		    			}
		    		}
		    	}
		    }

		    int other_player = 1 - player;

		    if (possible.empty()) {
		        best_result[player][state] = calc_result(other_player);  
		    } else if (waiting_counts[player] == 1) {
		        waiting_counts[player] -= 1;
		        best_result[player][state] = waiting_counts[1] - waiting_counts[0];
		        waiting_counts[player] += 1;
		    } else {
		        best_result[player][state] = hand_size * (player == 0 ? -1 : 1);

		        for (auto& card_index : possible) {
		        	auto& [value, mode] = all_cards[card_index];
		        
		        	auto mode_segment = segments[mode];
		        	
		        	auto& [left, right] = mode_segment;
		            	segments[mode] = make_pair(
		            		min(left, value),
		            		max(right, value)
            			);

		            	waiting_counts[player] -= 1;
		            	int result = calc_result(other_player);
		            	waiting_counts[player] += 1;

		            	segments[mode] = mode_segment;

			        best_result[player][state] = (player == 0) 
			        	? max(best_result[player][state], result)
			        	: min(best_result[player][state], result);
			}
		    }	

		    return best_result[player][state];
		};
		
		int alice = 0, bob = 1;

	        int alice_result = calc_result(alice);
	        int bob_result = calc_result(bob);

        	return abs(alice_result - bob_result);
	};
	
	auto card_to_str = [&](ii& card) {
	        auto [value, mode] = card;
	        
	        string card_str(2, '?');
	        
	        card_str[0] = values[value + nine];
	        card_str[1] = modes[mode];
	        
	        return card_str;
	};

	unordered_map<int, vector<vector<string>>> results_to_hands;

	auto rng = std::default_random_engine {};
	
	while (results_to_hands.size() < need_size) {
		std::shuffle(std::begin(all_cards), std::end(all_cards), rng);

        	int result = get_ans();
	        if (!results_to_hands.count(result)) {	
	        	results_to_hands[result].resize(players_count);
	        	for (int card_index = 0; card_index < all_cards.size(); ++card_index) {
	        		results_to_hands[result][card_index / hand_size].push_back(
	        			card_to_str(all_cards[card_index])
	        		);
	        	}
	        }
	}
	
	for (auto [_, ans_hands] : results_to_hands) {
		for (auto ans_hand : ans_hands) {
			for (auto card : ans_hand) cout << card << " ";
			cout << "\n";
		}
		cout << "\n";
	}
}

int main() {
	ios::sync_with_stdio(false);
	cin.tie(nullptr);
	
	int need_size;
    	cin >> need_size;
    	
    	using namespace std::chrono;
    	
    	auto start = high_resolution_clock::now();
    	
    	solve_test(need_size);
    	
    	auto stop = high_resolution_clock::now();
    	auto duration = duration_cast<milliseconds>(stop - start);
 
 	int millis_total = duration.count();
 	int seconds = millis_total / 1000, millis = millis_total % 1000;
 	
	cout << "Time taken: " << seconds << " sec " << millis << " ms" << endl;
	
	return 0;
}
